flow/table.go

```diff
@@ -113,29 +120,280 @@ func updateTCPFlagTime(prevFlagTime int64, currFlagTime int64) int64 {
+		layer := (path & (uint64(C.LAYERS_PATH_MASK) << uint(i*C.LAYERS_PATH_SHIFT))) >> uint(i*C.LAYERS_PATH_SHIFT)
+		if layer == 0 {
+			continue
+		}
+		if notFirst {
+			layersPath += "/"
+		}
+		notFirst = true
+		switch layer {
+		case C.ETH_LAYER:
+			layersPath += "Ethernet"
+		case C.ARP_LAYER:
+			layersPath += "ARP"
+		case C.DOT1Q_LAYER:
+			layersPath += "Dot1Q"
+		case C.IP4_LAYER:
+			layersPath += "IPv4"
+		case C.IP6_LAYER:
+			layersPath += "IPv6"
+		case C.ICMP4_LAYER:
+			layersPath += "ICMPv4"
+		case C.ICMP6_LAYER:
+			layersPath += "ICMPv6"
+		case C.UDP_LAYER:
+			layersPath += "UDP"
+		case C.TCP_LAYER:
+			layersPath += "TCP"
+		case C.SCTP_LAYER:
+			layersPath += "SCTP"
+		case C.GRE_LAYER:
+			hasGRE = true
+			layersPath += "GRE"
+		default:
+			layersPath += "Unknown"
+		}
+	}
+	return layersPath, hasGRE
+}
+
+func (ft *Table) newFlowFromEBPF(ebpfFlow *EBPFFlow, key string) *Flow {
+
+	f := NewFlow()
+	f.Init(common.UnixMillis(ebpfFlow.start), ebpfFlow.probeNodeTID, UUIDs{})
+	f.Last = common.UnixMillis(ebpfFlow.last)
+
+	layersInfo := uint8(ebpfFlow.kernFlow.layers_info)
+
+	// LINK
+	if layersInfo&uint8(C.LINK_LAYER_INFO) > 0 {
+		linkA := C.GoBytes(unsafe.Pointer(&ebpfFlow.kernFlow.link_layer.mac_src[0]), C.ETH_ALEN)
+		linkB := C.GoBytes(unsafe.Pointer(&ebpfFlow.kernFlow.link_layer.mac_dst[0]), C.ETH_ALEN)
+		f.Link = &FlowLayer{
+			Protocol: FlowProtocol_ETHERNET,
+			A:        net.HardwareAddr(linkA).String(),
+			B:        net.HardwareAddr(linkB).String(),
+			ID:       int64(ebpfFlow.kernFlow.link_layer.id),
+		}
+	}
+
+	var hasGRE bool
+	f.LayersPath, hasGRE = kernLayersPath(ebpfFlow.kernFlow)
+
+	if hasGRE {
+		pathSplitAt := strings.Index(f.LayersPath, "/GRE/")
+		innerLayerPath := f.LayersPath[pathSplitAt+len("/GRE/"):]
+
+		parent := f
+		parent.LayersPath = f.LayersPath[:pathSplitAt] + "/GRE"
+
+		// NETWORK
+		if layersInfo&uint8(C.NETWORK_LAYER_INFO) > 0 {
+			protocol := uint16(ebpfFlow.kernFlow.network_layer_outer.protocol)
+
+			switch protocol {
+			case syscall.ETH_P_IP:
+				netA := C.GoBytes(unsafe.Pointer(&ebpfFlow.kernFlow.network_layer_outer.ip_src[12]), net.IPv4len)
+				netB := C.GoBytes(unsafe.Pointer(&ebpfFlow.kernFlow.network_layer_outer.ip_dst[12]), net.IPv4len)
+				parent.Network = &FlowLayer{
+					Protocol: FlowProtocol_IPV4,
+					A:        net.IP(netA).To4().String(),
+					B:        net.IP(netB).To4().String(),
+				}
+			case syscall.ETH_P_IPV6:
+				netA := C.GoBytes(unsafe.Pointer(&ebpfFlow.kernFlow.network_layer_outer.ip_src[0]), net.IPv6len)
+				netB := C.GoBytes(unsafe.Pointer(&ebpfFlow.kernFlow.network_layer_outer.ip_dst[0]), net.IPv6len)
+				parent.Network = &FlowLayer{
+					Protocol: FlowProtocol_IPV6,
+					A:        net.IP(netA).String(),
+					B:        net.IP(netB).String(),
+				}
+			}
+		}
+		pkey := kernFlowKeyOuter(ebpfFlow.kernFlow)
+		parent.UpdateUUID(pkey, Opts{LayerKeyMode: L3PreferedKeyMode})
+		ft.table.Add(pkey, parent)
```

Comment:
I would prefer to have either both flows Added in this function or both returned, but here the parent is inserted here and the other flow in the caller

Reply:
done